### ISSUE TITLE
Treat unchanging dictionaries with known keys as ad-hoc `TypedDict`s to avoid errors involving `**kwargs`

### ISSUE THREAD

**Issue by alexmojaki** [microsoft/pyright#7270]
This code:

```python
def func(*, known_param: str | None = None, **kwargs: Any):
    ...

func(**{'foo': 1})
```

produces:

```
error: Argument of type "int" cannot be assigned to parameter "known_param" of type "str | None" in function "func"
    Type "int" cannot be assigned to type "str | None"
      "int" is incompatible with "str"
      "int" is incompatible with "None" (reportGeneralTypeIssues)
```

I understand this has been explained in multiple places such as https://github.com/microsoft/pyright/issues/5351#issuecomment-1601094654

In terms of some of the workarounds mentioned, note that `func` represents the shape of a real public API in a library. Making `known_param` positional-only is not viable as there are several similar optional arguments. The signature of `**kwargs: Any` cannot change to a `TypedDict` because it really could be anything.

This means that we have a dilemma for our API. Either we:

1. Change `known_param` to have type `Any` and users no longer get errors if they pass `known_param=3`, OR
2. Leave it as is, and expect users to deal with false positive type errors when they use `**kwargs`.

What I don't understand is why pyright treats `{'foo': 1}` as an opaque `dict[str, int]` as if it has no idea what's inside. This code suggests that it can infer when a dictionary literal works as a TypedDict:

```python
from typing import TypedDict

class Point(TypedDict):
    x: int
    y: int

def draw_point(p: Point):
    ...

# This seems to imply that `{'x': 1, 'y': 2}` can be inferred as a Point
draw_point({'x': 1, 'y': 2})  # this is fine
draw_point({'X': 1, 'Y': 2})  # this is an error:
"""
error: Argument of type "dict[str, int]" cannot be assigned to parameter "p" of type "Point" in function "draw_point"
    "X" is an undefined field in type "Point"
    "Y" is an undefined field in type "Point" (reportGeneralTypeIssues)
"""

def draw_x_y(x: int, y: int, text: str = ''):
    ...

# But that inference doesn't work here
draw_x_y(**Point(x=1, y=2))  # this is fine
draw_x_y(**{'x': 1, 'y': 2})  # this is an error:
"""
error: Argument of type "int" cannot be assigned to parameter "text" of type "str" in function "draw_x_y"
    "int" is incompatible with "str" (reportGeneralTypeIssues)
"""
```

So my proposal is that pyright treats the `{'x': 1, 'y': 2}` in `draw_x_y(**{'x': 1, 'y': 2})` as a `Point(x=1, y=2)`, regardless of whether such a class `Point` already exists or not. Not everywhere it can, which would probably be a huge slowdown, but if it finds an apparent error that *might* be resolved by inventing an appropriate TypedDict then it should try that.

**Comment by erictraut** [comment]
I think I understand the problem you're trying to solve, but your proposed solution is problematic for a number of reasons.

A TypedDict is not type compatible with a dictionary, so inferring a TypedDict from a dictionary expression would generally be a bad idea. I understand that you're proposing to do this only under certain circumstances, but I fear that it would result in behaviors that are hard to predict, understand, and diagnose when it doesn't produce the intended results.

Pyright's inference rules for dictionaries are already more complicated than I'd like, so I'm reluctant to make them more complicated. 

This would also be a big deviation from the inference behavior for all other Python type checkers, which goes against the ongoing efforts to make type checkers more consistent.

For those reasons, I'm going to reject this enhancement request.

I'm guessing that you've already thought of this workaround, but in case you hadn't... There's no need to use a dictionary literal expression and then unpack the dictionary within a function call. You can use keyword arguments directly.

```python
draw_x_y(x=1, y=2)
```

It may be worth starting a conversation about this use case in the [Python typing forum](https://discuss.python.org/c/typing/32). This forum includes members of the broader Python typing community including the maintainers of all the major Python type checkers.

**Comment by alexmojaki** [comment]
> A TypedDict is not type compatible with a dictionary, so inferring a TypedDict from a dictionary expression would generally be a bad idea.

Then why does `draw_point({'x': 1, 'y': 2})` work?

> There's no need to use a dictionary literal expression and then unpack the dictionary within a function call. You can use keyword arguments directly.

Here is the equivalent of the real use cases where I've run into this, to explain why I'm using kwargs:

```python
kwargs = {
    'foo1': 'bar1',
    'foo2': 'bar2',
    'foo3': 'bar3',
    'foo4': 'bar4',
    'foo5': 'bar5',
    'foo6': 'bar6',
}

func(**kwargs)
func2(**kwargs)
```

```python
try:
    kwargs = {
        'foo1': bar1(),
        'foo2': bar2(),
        'foo3': bar3(),
        'foo4': bar4(),
        'foo5': bar5(),
        'foo6': bar6(),
    }
except Exception:
    kwargs = {}

func(**kwargs)
```

**Comment by erictraut** [comment]
> Then why does draw_point({'x': 1, 'y': 2}) work?

Because the normal type inference rules for dict expressions don't apply here. Pyright is using [bidirectional type inference](https://microsoft.github.io/pyright/#/type-inference?id=bidirectional-type-inference-expected-types) in this case.

In your real use case example above, you're assigning a dictionary to an unannotated variable `kwargs`. It would never be appropriate in this case for the inferred type to be a TypedDict. (I thought you were previously proposing to infer a TypedDict type _only_ in the special case where a dictionary expression was used with a dictionary unpack operator in a call expression.) There is no bidirectional type inference context when assigning to `kwargs`, so the [normal type inference rules for dictionary expressions](https://microsoft.github.io/pyright/#/type-inference?id=dictionary-expressions) apply here.

**Comment by alexmojaki** [comment]
Based on your response I discovered this:

```python
from typing import TypedDict

class Point(TypedDict):
    x: int
    y: int

def draw_point(p: Point):
    ...

draw_point({'x': 1, 'y': 2})  # ok

point = {'x': 1, 'y': 2}
draw_point(point)  # error
```

I don't see why extracting to a variable changes things. What's the reasoning?

What's the expected way to solve those use cases?

**Comment by erictraut** [comment]
Yes, this is the way type inference works. This is consistent with how other Python type checkers (like mypy) do it. It's also consistent with other languages like TypeScript.

The expected way to solve this is to supply a type annotation for the `point` variable. This provides the context for bidirectional type inference and therefore overrides the normal type inference rules for dictionary expressions.

```python
point: Point = {'x': 1, 'y': 2}
draw_point(point)  # no error
```

You can also call the `Point` constructor explicitly.

```python
point = Point({'x': 1, 'y': 2})
draw_point(point)  # no error
```

**Comment by alexmojaki** [comment]
In the second use case at least I understand that magically inferring some kind of TypedDict is already becoming messy since there's two declarations with different sets of keys. But these cases don't actually require a `TypedDict`, I just used that because I could see that there was at least some existing inference in place for them. But really It just needs to know that e.g. `{'x': 1, 'y': 2}` *doesn't* contain the key `'text'`, which doesn't seem like a lot to ask for. Most likely this would be implemented by inferring the full set of possible keys (`x` and `y`) although that's a stricter requirement.

However it seems that inferring that wouldn't be enough, since even specifying the keys explicitly isn't enough for it to figure out what happens when the dict is unpacked. Shouldn't this work?

```python
from typing import Literal


def draw_x_y(x: int, y: int, text: str = ''):
    ...

point: dict[Literal['x', 'y'], int] = {'x': 1, 'y': 2}
draw_x_y(**point)  # error, which is not expected, especially given the below
print(point['text'])  # error, which is expected
```

**Comment by erictraut** [comment]
This is all working as intended. This isn't a bug. It's also consistent with how other type checkers work.

**Comment by alexmojaki** [comment]
This doesn't give the same result in pyright and mypy:

```python
from typing import Literal


def draw_x_y(x: int, y: int):
    ...

point: dict[Literal['x', 'y'], int] = {'x': 1, 'y': 2}
draw_x_y(**point)
```
